packages/node-patches/src/fs.ts

```diff
@@ -0,0 +1,345 @@
+                  return cb(err);
+                }
+
+                if (isOutPath(str)) {
+                  // if it's an out link we have to return the original stat.
+                  return fs.stat(args[0], cb);
+                }
+                // its a symlink and its inside of the root.
+                cb(false, stats);
+              }
+            );
+          }
+        }
+      };
+    }
+    origLstat(...args);
+  };
+
+  //tslint:disable-next-line:no-any
+  fs.realpath = (...args: any[]) => {
+    const cb = args.length > 1 ? args[args.length - 1] : undefined;
+    if (cb) {
+      args[args.length - 1] = (err: Error, str: string) => {
+        if (err) return cb(err);
+        if (isOutPath(str)) {
+          cb(false, path.resolve(args[0]));
+        } else {
+          cb(false, str);
+        }
+      };
+    }
+    origRealpath(...args);
+  };
+
+  //tslint:disable-next-line:no-any
+  fs.readlink = (...args: any[]) => {
+    const cb = args.length > 1 ? args[args.length - 1] : undefined;
+    if (cb) {
+      args[args.length - 1] = (err: Error, str: string) => {
+        if (err) return cb(err);
+        if (isOutPath(str)) {
+          const e = new Error(
+            "EINVAL: invalid argument, readlink '" + args[0] + "'"
+          );
+          //tslint:disable-next-line:no-any
+          (e as any).code = 'EINVAL';
+          // if its not supposed to be a link we have to trigger an EINVAL error.
+          cb(e);
+        } else {
+          cb(false, str);
+        }
+      };
+    }
+    origReadlink(...args);
+  };
+
+  //tslint:disable-next-line:no-any
+  fs.lstatSync = (...args: any[]) => {
+    let stats = origLstatSync(...args);
+    if (stats.isSymbolicLink() && isOutPath(origRealpathSync(args[0]))) {
+      stats = fs.statSync(...args);
+    }
+    return stats;
+  };
+
+  //tslint:disable-next-line:no-any
+  fs.realpathSync = (...args: any[]) => {
+    const str = origRealpathSync(...args);
+    if (isOutPath(str)) {
+      return path.resolve(args[0]);
+    }
+    return str;
+  };
+
+  //tslint:disable-next-line:no-any
+  fs.readlinkSync = (...args: any[]) => {
+    const str = origReadlinkSync(...args);
+    if (isOutPath(str)) {
+      const e = new Error(
+        "EINVAL: invalid argument, readlink '" + args[0] + "'"
+      );
+      //tslint:disable-next-line:no-any
+      (e as any).code = 'EINVAL';
+      throw e;
+    }
+    return str;
+  };
+
+  //tslint:disable-next-line:no-any
+  fs.readdir = (...args: any[]) => {
+    const p = path.resolve(args[0]);
+
+    const cb = args[args.length - 1];
+    if (typeof cb !== 'function') {
+      // this will likely throw callback required error.
+      return origReaddir(...args);
+    }
+
+    args[args.length - 1] = (err: Error, result: Dirent[]) => {
+      if (err) return cb(err);
+      // user requested withFileTypes
+      if (result[0] && result[0].isSymbolicLink) {
+        Promise.all(result.map((v: Dirent) => handleDirent(p, v)))
```

Comment:
Does it make sense to use `util.callbackify` here? Right now it's possible that `cb` is called twice if it throws in the initial invocation.

Reply:
i onced all the callbacks. thanks for alling this pout